internal/server/iigointernal/legislature.go

```
@@ -0,0 +1,166 @@
+	//TODO: log of vote occurring with ruleID, clientIDs
+	//TODO: log of clientIDs vs islandsAllowedToVote
+	//TODO: log of ruleID vs s.RuleToVote
+	return ruleVote.GetBallotBox()
+}
+
+//Speaker declares a result of a vote (see spec to see conditions on what this means for a rule-abiding speaker)
+//Called by orchestration
+func (l *legislature) announceVotingResult() error {
+
+	rule, result, err := l.clientSpeaker.DecideAnnouncement(l.ruleToVote, l.votingResult)
+
+	if err == nil {
+		//Deduct action cost
+		l.budget -= serviceCharge
+
+		//Reset
+		l.ruleToVote = ""
+		l.votingResult = false
+
+		//Perform announcement
+		broadcastToAllIslands(shared.TeamIDs[l.SpeakerID], generateVotingResultMessage(rule, result))
+		return l.updateRules(rule, result)
+	}
+	return nil
+}
+
+func generateVotingResultMessage(ruleID string, result bool) map[int]baseclient.Communication {
+	returnMap := map[int]baseclient.Communication{}
+
+	returnMap[RuleName] = baseclient.Communication{
+		T:        baseclient.CommunicationString,
+		TextData: ruleID,
+	}
+	returnMap[RuleVoteResult] = baseclient.Communication{
+		T:           baseclient.CommunicationBool,
+		BooleanData: result,
+	}
+
+	return returnMap
+}
+
+//reset resets internal variables for safety
+func (l *legislature) reset() {
+	l.ruleToVote = ""
+	l.ballotBox = voting.BallotBox{}
+	l.votingResult = false
+}
+
+// updateRules updates the rules in play according to the result of a vote.
+func (l *legislature) updateRules(ruleName string, ruleVotedIn bool) error {
+	l.budget -= serviceCharge
+	//TODO: might want to log the errors as normal messages rather than completely ignoring them? But then Speaker needs access to client's logger
+	notInRulesCache := errors.Errorf("Rule '%v' is not available in rules cache", ruleName)
+	if ruleVotedIn {
+		// _ = rules.PullRuleIntoPlay(ruleName)
+		err := rules.PullRuleIntoPlay(ruleName)
+		if err != nil {
+			if err.Error() == notInRulesCache.Error() {
+				return err
+			}
```

Why?

---

Sorry, just looking at this now, I will overhaul how Rules returns errors in a new PR. I'll make an issue and resolve in time.

---

Sounds good.

---

> Why?

Given the current circumstances, PullRuleIntoPlay returns an error when a rule is pulled into play when it is already in play. updateRules ignores that error. But it does need to return an error when a rule is not in the rules cache. Therefore it's only returning the notInRulesCache error